Allow connectApp command to have dependencies

juan-cortes opened the pull request:
### What is this
This is an evolution of the `connectApp` logic that allows us to provide an optional array of `AppRequest` (can be accounts, currencies, or app names) that will be required in order to complete the command. In case there are such dependencies we will check the apps on the device and either throw an error or attempt to install them if some are missing, or continue if all requirements are met.

It's completely plug and play into existing flows and requires close to no changes on LLD, basically changing the way we display the error case to account for multiple app dependencies.
![image](https://user-images.githubusercontent.com/4631227/117148644-b78c2400-adb6-11eb-9646-53cf734e2c66.png)

### What can it be used for
For currently existing flows this enables us to remove the device wall on the swap feature and install all the missing dependencies but could potentially be used for all future features that can require plugins/apps on their flows.

### What is the device wall
The device wall is the current flow killer check we do when accessing the swap feature, if the user is missing a dependency app, or the exchange app itself we kill the user flow and prevent them from going further. In the case of the exchange app we do this even before the user has a chance to see what the feature is all about.

gre commented on src/apps/hw.js:
```diff
@@ -52,10 +51,15 @@ export type StreamAppInstallEvent =
-  appName: string
-): Observable<StreamAppInstallEvent | ConnectAppEvent> =>
+  appNames: string[],
+  maybeOb?: () => Observable<*>,
```
need more explicit naming.

gre commented on src/apps/hw.js:
```diff
@@ -70,20 +74,29 @@ export const streamAppInstall = (
+
+          if (
+            isOutOfMemoryState(predictOptimisticState(state)) ||
+            !getEnv("EXPERIMENTAL_INLINE_INSTALL")
```
i don't understand why the `if` is here.

juan-cortes replied:
Doing the listapps and install plan allows me to know what _would be_ installed even if I don't finally do it because the user didn't opt in for inline installation. I need a way of knowing what is the error case data if you get what I mean

gre commented on src/hw/actions/app.js:
```diff
@@ -46,7 +47,7 @@ type State = {|
   requestQuitApp: boolean,
   requestOpenApp: ?string,
-  requiresAppInstallation: ?{ appName: string },
+  requiresAppInstallation: ?{ appName: string } | ?{ appNames: string[] },
```
in one side you try not to be too breaking (splitting into two actions) but on the state itself, i feel it's actually breaking (if UI side expects .appName)
so i have feeling we should just move to `Array<*>` paradigm everywhere to avoid all the ifs.

juan-cortes replied:
Makes sense

gre commented on src/hw/requireApps.js:
```diff
@@ -0,0 +1,82 @@
+  appNames: string[],
+};
+
+const cmd = ({ devicePath, appNames }: Input): Observable<ConnectAppEvent> =>
```
I don't understand why this is not done in actions/app , yet you add to change code here.

i miss context of this PR to have any conclusion on this.

let's rediscuss next week.

juan-cortes replied:
That was the whole point of wanting you to take a look. I can rework it so it's in the same connectApp but it seemed odd because the end result is not "opening" an app. 🤔

juan-cortes committed:
Add optional dependencies to connectApp

In order to allow flows to have dependent application installed we introduce the dependencies array which will check (and attempt to install) the installed apps on the device before attempting to open the target one.

juan-cortes commented on src/apps/filtering.js:
```diff
@@ -30,11 +30,16 @@ type UpdateAwareInstalledApps = {
     currency ? `${currency.name} ${currency.ticker}` : ""
   }`;
-  return terms.toLowerCase().includes(query.toLowerCase().trim());
+  return queries.some((query) => terms.toLowerCase().includes(query));
```
Would drop this if we go the delete all route

juan-cortes commented on src/hw/connectApp.js:
```diff
@@ -154,19 +160,28 @@ const cmd = ({
+                return streamAppInstall({
+                  transport,
+                  appNames: [appName, ...dependencies],
+                  onSuccessObs: () => innerSub({ appName }), // NB without deps
```
Return the same thing we just did, but without the dependency array since we completed the inline install

gre commented on src/hw/actions/app.js:
```diff
@@ -263,7 +264,7 @@ const reducer = (state: State, e: Event): State => {
         allowManagerGranted: false,
         allowManagerRequestedWording: null,
-        requiresAppInstallation: { appName: e.appName },
+        requiresAppInstallation: { appNames: e.appNames },
```
this requires a change on UI, correct? should we temporarily still have a `appName: e.appNames[lastIndex]` ?

juan-cortes replied:
This would definitely need a minor adaptation of the UI, but it's aligned with the redesign that we discussed today with @cthepot-ledger, I think passing appName temporarily to not break at all can be fine if you deem it necessary

gre reviewed (commented):
minor concern regarding one possible breaking change

juan-cortes committed:
Make connectApp changes non breaking

Keep the event name and signature compatible with current implementation so that a transition towards multiple missing apps is not a breaking change.

juan-cortes committed:
Merge branch 'master' into requireAppAction

juan-cortes committed:
Autoquit if we have dependencies and we're not in dashboard

In order to perform the listapps that happens inside the streamAppInstall code, we need to be in the dashboard. There was an edge case where we have dependencies and the user is already in the target app, we would be skipping the dependency check. Nice catch @gre

gre commented on src/apps/hw.js:
```diff
@@ -70,20 +74,31 @@ export const streamAppInstall = (
+          let state = initState(e.result);
+          appNames.forEach(
+            (name) => (state = reducer(state, { type: "install", name }))
+          );
```
you can refactor this to:

```
const state = appNames.reduce((state, name) => reducer(state, { type: "install", name }), initState(e.result));
```

as it's a reducer :)

juan-cortes replied:
nice, nice

juan-cortes replied:
You got it

gre reviewed (commented):
i have one slight request to polish but to me we are good.

juan-cortes committed:
refactor state for stream install apps

juan-cortes committed:
Merge branch 'master' into requireAppAction

gre committed:
Merge branch 'master' into requireAppAction
